File: .sling/libs/web5/eqela/jk-data/0.32.0/jk.diff.js
let JkDiffUnifiedDiff = function() {};

JkDiffUnifiedDiff.prototype._t = { "JkDiffUnifiedDiff" : true };
JkDiffUnifiedDiff.prototype._tobj = JkDiffUnifiedDiff;

JkDiffUnifiedDiff.NEW = function() {
	var v = new JkDiffUnifiedDiff;
	return v.CTOR_JkDiffUnifiedDiff();
};

JkDiffUnifiedDiff.prototype.CTOR_JkDiffUnifiedDiff = function() {
	return this;
};

JkDiffUnifiedDiff.backtrack = function(lcs, olines, nlines, i, j) {
	if(i == 0 || j == 0) {
		return new Array;
	}
	var ocount = JkLangVector.getSize(olines);
	if(olines[i - 1] == nlines[j - 1]) {
		var v = JkDiffUnifiedDiff.backtrack(lcs, olines, nlines, (i - 1), (j - 1));
		v.push(olines[i - 1]);
		return v;
	}
	if(lcs[i + (j - 1) * (ocount + 1)] > lcs[i - 1 + j * (ocount + 1)]) {
		return JkDiffUnifiedDiff.backtrack(lcs, olines, nlines, i, (j - 1));
	}
	return JkDiffUnifiedDiff.backtrack(lcs, olines, nlines, (i - 1), j);
};

JkDiffUnifiedDiff.execute = function(fileName, original, modified) {
	if(!(original != null)) {
		return null;
	}
	if(!(modified != null)) {
		return null;
	}
	var olines = JkLangString.split(original, '\n'.charCodeAt(), 0);
	var nlines = JkLangString.split(modified, '\n'.charCodeAt(), 0);
	var ocount = JkLangVector.getSize(olines);
	var ncount = JkLangVector.getSize(nlines);
	var lcs = [];
	for(var n = 0 ; n < ocount + 1 ; n++) {
		lcs[n] = 0;
	}
	for(var n1 = 0 ; n1 < ncount + 1 ; n1++) {
		lcs[n1 * (ocount + 1)] = 0;
	}
	for(var i = 1 ; i < ocount + 1 ; i++) {
		for(var j = 1 ; j < ncount + 1 ; j++) {
			if(olines[i - 1] == nlines[j - 1]) {
				lcs[i + j * (ocount + 1)] = lcs[i - 1 + (j - 1) * (ocount + 1)] + 1;
			}
			else if(lcs[i + (j - 1) * (ocount + 1)] > lcs[i - 1 + j * (ocount + 1)]) {
				lcs[i + j * (ocount + 1)] = lcs[i + (j - 1) * (ocount + 1)];
			}
			else {
				lcs[i + j * (ocount + 1)] = lcs[i - 1 + j * (ocount + 1)];
			}
		}
	}
	var lcsLines = JkDiffUnifiedDiff.backtrack(lcs, olines, nlines, ocount, ncount);
	var op = 0;
	var np = 0;
	var sameCounter = 0;
	var blocksb = null;
	var blockopc = 0;
	var blocknpc = 0;
	var blockolines = 0;
	var blocknlines = 0;
	var msb = JkLangStringBuilder.NEW();
	msb.appendString("diff -u a/" + (JkLangString.safeString(fileName)) + (" b/") + (JkLangString.safeString(fileName)) + ("\n"));
	var timeNow = JkTimeSystemClock.asDateTimeLocal();
	var timeString = (JkLangString.safeString((timeNow.toStringDateTime()))) + (".000000000 +0000");
	msb.appendString("--- a/" + (JkLangString.safeString(fileName)) + (" ") + (JkLangString.safeString(timeString)) + ("\n"));
	msb.appendString("+++ b/" + (JkLangString.safeString(fileName)) + (" ") + (JkLangString.safeString(timeString)) + ("\n"));
	var hdr = msb.toString();
	msb = JkLangStringBuilder.NEW();
	var lp = 0;
	var lcsLinesCount = JkLangVector.getSize(lcsLines);
	while(true){
		var line = null;
		if(lp < lcsLinesCount) {
			line = lcsLines[lp];
		}
		else {
			line = null;
		}
		lp++;
		if(line == null && op >= ocount && np >= ncount) {
			break;
		}
		var changes = null;
		var oop = op;
		var onp = np;
		while(op < ocount && !(olines[op] == line)){
			if(!(changes != null)) {
				changes = new Array;
			}
			changes.push("-" + (JkLangString.safeString(olines[op])));
			op++;
		}
		while(np < ncount && !(nlines[np] == line)){
			if(!(changes != null)) {
				changes = new Array;
			}
			changes.push("+" + (JkLangString.safeString(nlines[np])));
			np++;
		}
		if(changes != null) {
			sameCounter = 0;
			if(!(blocksb != null)) {
				var opc = oop - 3;
				if(opc < 0) {
					opc = 0;
				}
				var npc = onp - 3;
				if(npc < 0) {
					npc = 0;
				}
				blocksb = JkLangStringBuilder.NEW();
				blockopc = opc + 1;
				blocknpc = npc + 1;
				blockolines = 0;
				blocknlines = 0;
				for(var n2 = opc ; n2 < oop ; n2++) {
					blocksb.appendString(" " + (JkLangString.safeString(olines[n2])) + ("\n"));
					blockolines++;
				}
			}
			if(changes != null) {
				var n21 = 0;
				var m = changes.length;
				for(n21 = 0 ; n21 < m ; n21++) {
					var change = changes[n21];
					if(change != null) {
						blocksb.appendString(change);
						blocksb.appendCharacter('\n'.charCodeAt());
						if(JkLangString.startsWith(change, "+", 0)) {
							blocknlines++;
						}
						else if(JkLangString.startsWith(change, "-", 0)) {
							blocknlines--;
							blockolines++;
						}
					}
				}
			}
			changes = null;
		}
		sameCounter++;
		if(sameCounter > 3) {
			if(blocksb != null) {
				if(hdr != null) {
					msb.appendString(hdr);
					hdr = null;
				}
				msb.appendString("@@ -" + (JkLangString.forInteger(blockopc)) + (",") + (JkLangString.forInteger(blockolines)) + (" +") + (JkLangString.forInteger(blocknpc)) + (",") + (JkLangString.forInteger((blockolines + blocknlines))) + (" @@\n"));
				msb.appendString((blocksb.toString()));
				blocksb = null;
			}
		}
		if(blocksb != null && op < ocount) {
			blocksb.appendString(" " + (JkLangString.safeString(olines[op])) + ("\n"));
			blockolines++;
		}
		op++;
		np++;
	}
	if(blocksb != null) {
		if(hdr != null) {
			msb.appendString(hdr);
			hdr = null;
		}
		msb.appendString("@@ -" + (JkLangString.forInteger(blockopc)) + (",") + (JkLangString.forInteger(blockolines)) + (" +") + (JkLangString.forInteger(blocknpc)) + (",") + (JkLangString.forInteger((blockolines + blocknlines))) + (" @@\n"));
		msb.appendString((blocksb.toString()));
		blocksb = null;
	}
	return msb.toString();
};

JkDiffUnifiedDiff.IS_INSTANCE = function(o) {
	return o != null && o._t != null && o._t["JkDiffUnifiedDiff"] === true;
};

File: .sling/libs/sushi/eqela/jk-data/0.23.0/jk.diff.lua
jk = jk or {}

jk.diff = jk.diff or {}

jk.diff.UnifiedDiff = {}
jk.diff.UnifiedDiff.__index = jk.diff.UnifiedDiff
_vm:set_metatable(jk.diff.UnifiedDiff, {})

function jk.diff.UnifiedDiff._create()
	local v = _vm:set_metatable({}, jk.diff.UnifiedDiff)
	return v
end

function jk.diff.UnifiedDiff:_init()
	self._isClassInstance = true
	self._qualifiedClassName = self._qualifiedClassName or 'jk.diff.UnifiedDiff'
	self['_isType.jk.diff.UnifiedDiff'] = true
end

function jk.diff.UnifiedDiff:_construct0()
	jk.diff.UnifiedDiff._init(self)
	return self
end

function jk.diff.UnifiedDiff:backtrack(lcs, olines, nlines, i, j)
	if i == 0 or j == 0 then
		do return {} end
	end
	do
		local ocount = _g.jk.lang.Vector:getSize(olines)
		if olines[i - 1 + 1] == nlines[j - 1 + 1] then
			local v = _g.jk.diff.UnifiedDiff:backtrack(lcs, olines, nlines, i - 1, j - 1)
			do _g.jk.lang.Vector:append(v, olines[i - 1 + 1]) end
			do return v end
		end
		if lcs[i + (j - 1) * (ocount + 1) + 1] > lcs[i - 1 + j * (ocount + 1) + 1] then
			do return _g.jk.diff.UnifiedDiff:backtrack(lcs, olines, nlines, i, j - 1) end
		end
		do return _g.jk.diff.UnifiedDiff:backtrack(lcs, olines, nlines, i - 1, j) end
	end
end

function jk.diff.UnifiedDiff:execute(fileName, original, modified)
	if not (original ~= nil) then
		do return nil end
	end
	if not (modified ~= nil) then
		do return nil end
	end
	do
		local olines = _g.jk.lang.String:split(original, 10, 0)
		local nlines = _g.jk.lang.String:split(modified, 10, 0)
		local ocount = _g.jk.lang.Vector:getSize(olines)
		local ncount = _g.jk.lang.Vector:getSize(nlines)
		local lcs = _vm:allocate_array((ocount + 1) * (ncount + 1))
		do
			local n = 0
			while n < ocount + 1 do
				lcs[n + 1] = 0
				do n = n + 1 end
			end
		end
		do
			local n = 0
			while n < ncount + 1 do
				lcs[n * (ocount + 1) + 1] = 0
				do n = n + 1 end
			end
		end
		do
			local i = 1
			while i < ocount + 1 do
				do
					local j = 1
					while j < ncount + 1 do
						if olines[i - 1 + 1] == nlines[j - 1 + 1] then
							lcs[i + j * (ocount + 1) + 1] = lcs[i - 1 + (j - 1) * (ocount + 1) + 1] + 1
						elseif lcs[i + (j - 1) * (ocount + 1) + 1] > lcs[i - 1 + j * (ocount + 1) + 1] then
							lcs[i + j * (ocount + 1) + 1] = lcs[i + (j - 1) * (ocount + 1) + 1]
						else
							lcs[i + j * (ocount + 1) + 1] = lcs[i - 1 + j * (ocount + 1) + 1]
						end
						do j = j + 1 end
					end
				end
				do i = i + 1 end
			end
		end
		do
			local lcsLines = _g.jk.diff.UnifiedDiff:backtrack(lcs, olines, nlines, ocount, ncount)
			local op = 0
			local np = 0
			local sameCounter = 0
			local blocksb = nil
			local blockopc = 0
			local blocknpc = 0
			local blockolines = 0
			local blocknlines = 0
			local msb = _g.jk.lang.StringBuilder._construct0(_g.jk.lang.StringBuilder._create())
			do msb:appendString("diff -u a/" .. _g.jk.lang.String:safeString(fileName) .. " b/" .. _g.jk.lang.String:safeString(fileName) .. "\n") end
			do
				local timeNow = _g.jk.time.SystemClock:asDateTimeLocal()
				local timeString = _g.jk.lang.String:safeString(timeNow:toStringDateTime()) .. ".000000000 +0000"
				do msb:appendString("--- a/" .. _g.jk.lang.String:safeString(fileName) .. " " .. _g.jk.lang.String:safeString(timeString) .. "\n") end
				do msb:appendString("+++ b/" .. _g.jk.lang.String:safeString(fileName) .. " " .. _g.jk.lang.String:safeString(timeString) .. "\n") end
				do
					local hdr = msb:toString()
					msb = _g.jk.lang.StringBuilder._construct0(_g.jk.lang.StringBuilder._create())
					do
						local lp = 0
						local lcsLinesCount = _g.jk.lang.Vector:getSize(lcsLines)
						while true do
							local line = nil
							if lp < lcsLinesCount then
								line = lcsLines[lp + 1]
							else
								line = nil
							end
							do lp = lp + 1 end
							if line == nil and op >= ocount and np >= ncount then
								do break end
							end
							do
								local changes = nil
								local oop = op
								local onp = np
								while op < ocount and not (olines[op + 1] == line) do
									if not (changes ~= nil) then
										changes = {}
									end
									do _g.jk.lang.Vector:append(changes, "-" .. _g.jk.lang.String:safeString(olines[op + 1])) end
									do op = op + 1 end
								end
								while np < ncount and not (nlines[np + 1] == line) do
									if not (changes ~= nil) then
										changes = {}
									end
									do _g.jk.lang.Vector:append(changes, "+" .. _g.jk.lang.String:safeString(nlines[np + 1])) end
									do np = np + 1 end
								end
								if changes ~= nil then
									sameCounter = 0
									if not (blocksb ~= nil) then
										local opc = oop - 3
										if opc < 0 then
											opc = 0
										end
										do
											local npc = onp - 3
											if npc < 0 then
												npc = 0
											end
											blocksb = _g.jk.lang.StringBuilder._construct0(_g.jk.lang.StringBuilder._create())
											blockopc = opc + 1
											blocknpc = npc + 1
											blockolines = 0
											blocknlines = 0
											do
												local n = opc
												while n < oop do
													do blocksb:appendString(" " .. _g.jk.lang.String:safeString(olines[n + 1]) .. "\n") end
													do blockolines = blockolines + 1 end
													do n = n + 1 end
												end
											end
										end
									end
									if changes ~= nil then
										local n2 = 0
										local m = _g.jk.lang.Vector:getSize(changes)
										do
											n2 = 0
											while n2 < m do
												local change = changes[n2 + 1]
												if change ~= nil then
													do blocksb:appendString(change) end
													do blocksb:appendCharacter(10) end
													if _g.jk.lang.String:startsWith(change, "+", 0) then
														do blocknlines = blocknlines + 1 end
													elseif _g.jk.lang.String:startsWith(change, "-", 0) then
														do blocknlines = blocknlines - 1 end
														do blockolines = blockolines + 1 end
													end
												end
												do n2 = n2 + 1 end
											end
										end
									end
									changes = nil
								end
								do sameCounter = sameCounter + 1 end
								if sameCounter > 3 then
									if blocksb ~= nil then
										if hdr ~= nil then
											do msb:appendString(hdr) end
											hdr = nil
										end
										do msb:appendString("@@ -" .. _g.jk.lang.String:forInteger(blockopc) .. "," .. _g.jk.lang.String:forInteger(blockolines) .. " +" .. _g.jk.lang.String:forInteger(blocknpc) .. "," .. _g.jk.lang.String:forInteger(blockolines + blocknlines) .. " @@\n") end
										do msb:appendString(blocksb:toString()) end
										blocksb = nil
									end
								end
								if blocksb ~= nil and op < ocount then
									do blocksb:appendString(" " .. _g.jk.lang.String:safeString(olines[op + 1]) .. "\n") end
									do blockolines = blockolines + 1 end
								end
								do op = op + 1 end
								do np = np + 1 end
							end
						end
						if blocksb ~= nil then
							if hdr ~= nil then
								do msb:appendString(hdr) end
								hdr = nil
							end
							do msb:appendString("@@ -" .. _g.jk.lang.String:forInteger(blockopc) .. "," .. _g.jk.lang.String:forInteger(blockolines) .. " +" .. _g.jk.lang.String:forInteger(blocknpc) .. "," .. _g.jk.lang.String:forInteger(blockolines + blocknlines) .. " @@\n") end
							do msb:appendString(blocksb:toString()) end
							blocksb = nil
						end
						do return msb:toString() end
					end
				end
			end
		end
	end
end
